### PR TITLE
TGUI Long Range Teleporter [MDB IGNORE]

### PR DESCRIPTION
**Pull request by Azrun**
[balance]
## About the PR 
Adds TGUI for LRT.
image
image

Allows silicon folks to use LRT and Quantum Telescope at range.

## Why's this needed? 
I want this for Artemis...

## Changelog

```changelog
(u)Azrun
(+)Silicon folks can now use the Long Range Teleporter and Quantum Telescope from a distance.
```

### CODE REVIEW

**Review by ZeWaka — APPROVED** [review]

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/LongRangeTeleporter.js:24]: "ref" is null for a subset of destinations, which would mean the same key for them. Consider `${ref ?? name}` or some compound key (i.e. use both).

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/LongRangeTeleporter.js:36]:
```suggestion
                    Receive
```

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/LongRangeTeleporter.js:29]:
```suggestion
                    Send
```

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (Azrun) [reply code/WorkInProgress/telescope/telescopeMisc.dm:178]: Checked in is_good_location() further down call tree. Would you prefer earlier validation? Right not it covers tgui, legacy, and mechcomp with current implementation.

Comment (mordent-goonstation) [reply code/WorkInProgress/telescope/telescopeMisc.dm:178]: Fine as long as it's checked!

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (mordent-goonstation) [inline code/WorkInProgress/telescope/telescopeMisc.dm:29]: I would think that the rest of this proc is probably not needed now? If it is, why?

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (mordent-goonstation) [inline code/WorkInProgress/telescope/telescopeMisc.dm:160]: Might be missing something, but I don't see where any destination item has a non-null `ref`?

File: maps/unused/oshan_xmas_2018.dmm
Comment (mordent-goonstation) [inline maps/unused/oshan_xmas_2018.dmm:10783]: All of these blank-line changes seem unneeded?

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (Azrun) [reply code/WorkInProgress/telescope/telescopeMisc.dm:29]: If Tgui were to fail it would use old UI. Can remove. 🙂

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (Azrun) [reply code/WorkInProgress/telescope/telescopeMisc.dm:160]: You are not. Ref is intended for use with Artemis planetary teleporter, to allow for decoupling from special places.

File: maps/unused/oshan_xmas_2018.dmm
Comment (Azrun) [reply maps/unused/oshan_xmas_2018.dmm:10783]: I can not speak to operations that were performed on unmaintained maps by automated mechanisms.... 🤖 
Owned by githook. 💀

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (Azrun) [reply tgui/packages/tgui/interfaces/LongRangeTeleporter.js:24]: Compound it is!

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (Azrun) [reply code/WorkInProgress/telescope/telescopeMisc.dm:29]: Removing!

File: maps/unused/oshan_xmas_2018.dmm
Comment (Azrun) [reply maps/unused/oshan_xmas_2018.dmm:10783]: Do you want me to remove these changes?

File: maps/manta_damaged.dmm
Comment (ZeWaka) [inline maps/manta_damaged.dmm:626]: i'll fix theses once merged ping me

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/LongRangeTeleporter.js:24]:
```suggestion
              <LabeledList.Item label={d["destination"]} key={`${d["destination"]}}>
```
If "destination" is not unique then the user has no way of differentiating between the locations, so destination *should* be enough to uniquely identify the `Item` by.

If for some reason `ref` is needed as well, note that casting it to string will be `"[Object object]"` in cases where `ref` is null.

You could do the below instead:

```suggestion
              <LabeledList.Item label={d["destination"]} key={`${d["destination"]}${d["ref"] ? `-${d["ref"]}` : ''}}>
```

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/LongRangeTeleporter.js:28]:
```suggestion
                    onClick={() => act("send", { target: d["ref"], name: d["destination"] })}
                  >
```

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/LongRangeTeleporter.js:35]:
```suggestion
                    onClick={() => act("receive", { target: d["ref"], name: d["destination"] })}
                  >
```
Pedantic code styling. Could also have it same-line, but the space before the `>` is *weird*.

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/LongRangeTeleporter.js:43]:
```suggestion
              </LabeledList.Item>
            )}
```
Pedantic code styling.

**Review by mordent-goonstation — APPROVED** [review]
Approving assuming below changes are made.

File: tgui/packages/tgui/interfaces/LongRangeTeleporter.js
Comment (Azrun) [reply tgui/packages/tgui/interfaces/LongRangeTeleporter.js:24]: Good catch, the implementation is myopic to the intended cases and ignorant of the possible user interaction aspect. ☹️

Comment (mordent-goonstation) [reply tgui/packages/tgui/interfaces/LongRangeTeleporter.js:24]: @Azrun Updated the suggestion to be (I think) syntactically correct. I was missing a closing backtick in my initial version.

**Comment by AffectedArc07** [comment]
Please put `[MDB IGNORE]` into your PR title or your repo will be blacklisted from MapDiffBot

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (mordent-goonstation) [inline code/WorkInProgress/telescope/telescopeMisc.dm:133]: TL;DR at bottom of comment.

This *should* be covered by `ui_act` on `datum` (see [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/modules/tgui/external.dm#L78)).

It returns `1` if the `ui.status` is not `UI_INTERACTIVE`, which is why the `if (.) return` bit is done above.

`ui.status` is generated from the `ui_status` proc, which exists on `/obj/machinery` [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/obj/machinery.dm#L186). The `state` it uses is determined the `ui_state` proc, which exists on `/obj/machinery` [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/obj/machinery.dm#L180).

In your case, it doesn't require physical access, so the `tgui_default_state` is used. That's defined [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/modules/tgui/states/default.dm).

The assorted stuff in there determines whether the UI should close, be open but not updating, open and updating but not allowing inputs, or be fully interactive (UI_INTERACTIVE).

Each type of mob handles that a little differently, and that's sort of the point (i.e. whether a human should be able to act is different from whether the AI should be able to).

Specifically for humans, it's only interactive if:
* they have a client connected, they are conscious, they are not unable to act for a few reasons (e.g. cuffs) - all of this is covered by `shared_ui_interaction`.
* the thing is not in a storage item, has the result of `get_dist` be `<= 1` (this does *not* deal with the z-level bit), the thing is in view (this *should* cover the z-level bit, I think?), is in interact range (uses `in_interact_range`) - all of this is covered by `shared_living_ui_distance`.

It's a little different for cyborgs/AI, but the key is that "you don't have to worry about it, it's covered by this stuff".

It *would* be worth sanity checking that the `in view` bit of [this](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/modules/tgui/states.dm#L117) does account for z-levels. If it does, you can remove the line this comment is attached to. If it does not, we should think about solving that.

I appreciate that this is an in-depth comment, meant to educate about the whole thing. TL;DR: this line *shouldn't* be needed, if it is then we should change something else instead.

**Comment by AffectedArc07** [comment]
Much obliged

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]
One last comment, this is getting pretty in-depth into tgui but I would like to use this PR as a good tgui example so holding it to a high standard. This *should* be the last thing, sorry for not catching earlier.

File: maps/manta_damaged.dmm
Comment (mordent-goonstation) [reply maps/manta_damaged.dmm:626]: Why not fix before merge? 🤔

File: code/WorkInProgress/telescope/telescopeMisc.dm
Comment (Azrun) [reply code/WorkInProgress/telescope/telescopeMisc.dm:133]: Added #5683 to capture desired behavior.

**Review by mordent-goonstation — APPROVED** [review]

**Pull request merged**
